### PR TITLE
Implement DependentMemberTypeRef support for the typeref->demangle tr…

### PR DESCRIPTION
**Pull request by tbkka**
…anslation

* Risk: Without this, RemoteMirror (and hence client apps) crash on certain inputs
* Resolves: rdar://108809238
* Testing:  Ran a client app with a custom-built library to verify it no longer crashes.  I've not been able to figure out how to build a unit test that hits this specific code path (in all my unit tests, dependent members seem to get fully substituted before we get to this section of code).

### CODE REVIEW

**Comment by tbkka** [comment]
@swift-ci Please test

File: stdlib/public/RemoteInspection/TypeRef.cpp [hunk 1]
```diff
@@ -852,11 +852,28 @@ class DemanglingForTypeRef
...
+      // Otherwise, build up a DependentAssociatedTR node with
+      // the member Identifer and protocol
+      auto AssocTy = Dem.createNode(Node::Kind::DependentAssociatedTypeRef);
+      AssocTy->addChild(MemberId, Dem);
+      auto Proto = Dem.demangleType(MangledProtocol);
+      assert(Proto && "Failed to demangle");
+      assert(Proto->getKind() == Node::Kind::Type && "Protocol type is not a type?!");
+      AssocTy->addChild(Proto, Dem);
+      node->addChild(AssocTy, Dem);
+    }
```
[the remark this answers is not in the document]
Comment (al45tair) [reply stdlib/public/RemoteInspection/TypeRef.cpp:876]: That looks right to me.

**Review by al45tair — APPROVED** [review]

File: stdlib/public/RemoteInspection/TypeRef.cpp [hunk 2]
```diff
@@ -852,11 +852,28 @@ class DemanglingForTypeRef
...
+    auto MangledProtocol = DM->getProtocol();
+    if (MangledProtocol.empty()) {
+      // If there's no protocol, add the Member as an Identifier node
+      node->addChild(MemberId, Dem);
+    } else {
+      // Otherwise, build up a DependentAssociatedTR node with
+      // the member Identifer and protocol
+      auto AssocTy = Dem.createNode(Node::Kind::DependentAssociatedTypeRef);
+      AssocTy->addChild(MemberId, Dem);
+      auto Proto = Dem.demangleType(MangledProtocol);
```
[the remark this answers is not in the document]
Comment (mikeash) [reply stdlib/public/RemoteInspection/TypeRef.cpp:871]: No, this looks fine as-is. If it was a temporary local use of the demangler then you might want to push and pop the node factory around this, but that's not what you're doing here. The node you allocate here will be cleared along with all the other nodes when some calling code pops the node factory.

File: stdlib/public/RemoteInspection/TypeRef.cpp [hunk 3]
```diff
@@ -852,11 +852,28 @@ class DemanglingForTypeRef
 
   Demangle::NodePointer
   visitDependentMemberTypeRef(const DependentMemberTypeRef *DM) {
-    assert(DM->getProtocol().empty() && "not implemented");
+
```
Comment (mikeash) [inline stdlib/public/RemoteInspection/TypeRef.cpp:855]: Stray newline?

File: stdlib/public/RemoteInspection/TypeRef.cpp [hunk 4]
```diff
@@ -852,11 +852,28 @@ class DemanglingForTypeRef
...
+    if (MangledProtocol.empty()) {
+      // If there's no protocol, add the Member as an Identifier node
+      node->addChild(MemberId, Dem);
+    } else {
+      // Otherwise, build up a DependentAssociatedTR node with
+      // the member Identifer and protocol
+      auto AssocTy = Dem.createNode(Node::Kind::DependentAssociatedTypeRef);
+      AssocTy->addChild(MemberId, Dem);
+      auto Proto = Dem.demangleType(MangledProtocol);
+      assert(Proto && "Failed to demangle");
```
Comment (mikeash) [inline stdlib/public/RemoteInspection/TypeRef.cpp:872]: Should we do a NULL check with some sort of graceful failure in no-asserts builds?

**Pull request merged**